File: apps/backend/analytics/compute.py
"""Orchestrator: run all pipelines for every user and upsert results."""

from __future__ import annotations

import logging

from db import get_conn
from insights import e1rm, frequency, summary, volume
from ml import forecast, plateau, recommendation
from queries import fetch_all_users, fetch_workout_records, upsert_insight

log = logging.getLogger(__name__)


async def run_all() -> dict:
    """Run all compute pipelines. Returns a summary dict."""
    async with get_conn() as conn:
        users = await fetch_all_users(conn)

    log.info("Starting compute for %d users", len(users))
    processed = 0
    errors = 0

    for user in users:
        try:
            await _compute_user(user.user_id)
            processed += 1
        except Exception:
            log.exception("Compute failed for user %s", user.user_id)
            errors += 1

    log.info("Compute done: %d processed, %d errors", processed, errors)
    return {"processed": processed, "errors": errors}


async def _compute_user(user_id: str) -> None:
    async with get_conn() as conn:
        records = await fetch_workout_records(conn, user_id)

    if not records:
        return

    async with get_conn() as conn:
        # Volume trend (aggregate, no exercise_id)
        vol_payload = volume.compute(records)
        if vol_payload:
            await upsert_insight(conn, user_id, "volume_trend", None, vol_payload)

        # Frequency (aggregate)
        freq_payload = frequency.compute(records)
        if freq_payload:
            await upsert_insight(conn, user_id, "frequency", None, freq_payload)

        # e1RM per exercise
        e1rm_map = e1rm.compute_per_exercise(records)
        for exercise_id, payload in e1rm_map.items():
            await upsert_insight(conn, user_id, "e1rm_progression", exercise_id, payload)

        # Exercise summary per slot
        summary_map = summary.compute_per_exercise(records)
        for exercise_id, payload in summary_map.items():
            await upsert_insight(conn, user_id, "exercise_summary", exercise_id, payload)

        for exercise_id, payload in plateau.compute_per_exercise(records).items():
            await upsert_insight(conn, user_id, "plateau_detection", exercise_id, payload)

        for exercise_id, payload in forecast.compute_per_exercise(records).items():
            await upsert_insight(conn, user_id, "e1rm_forecast", exercise_id, payload)

        for exercise_id, payload in recommendation.compute_per_exercise(records).items():
            await upsert_insight(conn, user_id, "load_recommendation", exercise_id, payload)

        await conn.commit()

File: apps/backend/api/src/analytics/compute.ts
/**
 * Analytics compute orchestrator.
 *
 * For each user it takes a transactional per-user lease, reads a bounded
 * freshness-consistent history, runs all seven pipelines, and atomically
 * replaces the insight snapshot. Upserts make retries idempotent.
 */

import { logger } from '../lib/logger';
import {
  fetchAllUsers,
  fetchWorkoutRecords,
  deleteComputedInsights,
  upsertInsight,
  withInsightTransaction,
  META_INSIGHT_TYPE,
} from './queries';
import { computeVolume } from './pipelines/volume';
import { computeFrequency } from './pipelines/frequency';
import { computeE1rmPerExercise } from './pipelines/e1rm';
import { computeSummaryPerExercise } from './pipelines/summary';
import { computePlateauPerExercise } from './pipelines/plateau';
import { computeForecastPerExercise } from './pipelines/forecast';
import { computeRecommendationPerExercise } from './pipelines/recommendation';

export interface RunAllSummary {
  readonly processed: number;
  readonly errors: number;
}

/** Run all pipelines for one user under the transaction/lease. */
export async function computeUser(userId: string): Promise<void> {
  await withInsightTransaction(userId, async (tx) => {
    // The transaction owns the per-user lease and the same user-row lock taken
    // by result/import/delete mutations. The bounded read and replacement are
    // therefore one freshness-consistent snapshot.
    const records = await fetchWorkoutRecords(userId, tx);
    await deleteComputedInsights(userId, tx);

    if (records.length > 0) {
      const volume = computeVolume(records);
      if (volume !== null) await upsertInsight(userId, 'volume_trend', null, volume, tx);

      const frequency = computeFrequency(records);
      if (frequency !== null) await upsertInsight(userId, 'frequency', null, frequency, tx);

      for (const [exerciseId, payload] of computeE1rmPerExercise(records)) {
        await upsertInsight(userId, 'e1rm_progression', exerciseId, payload, tx);
      }
      for (const [exerciseId, payload] of computeSummaryPerExercise(records)) {
        await upsertInsight(userId, 'exercise_summary', exerciseId, payload, tx);
      }
      for (const [exerciseId, payload] of computePlateauPerExercise(records)) {
        await upsertInsight(userId, 'plateau_detection', exerciseId, payload, tx);
      }
      for (const [exerciseId, payload] of computeForecastPerExercise(records)) {
        await upsertInsight(userId, 'e1rm_forecast', exerciseId, payload, tx);
      }
      for (const [exerciseId, payload] of computeRecommendationPerExercise(records)) {
        await upsertInsight(userId, 'load_recommendation', exerciseId, payload, tx);
      }
    }

    // Advance the fairness cursor only when the complete snapshot commits.
    await upsertInsight(userId, META_INSIGHT_TYPE, null, {}, tx);
  });
}

/** Run the compute for every eligible user. Returns a processed/errors summary. */
export async function runAll(): Promise<RunAllSummary> {
  const users = await fetchAllUsers();
  logger.info({ userCount: users.length }, 'analytics: starting compute');

  let processed = 0;
  let errors = 0;
  for (const user of users) {
    try {
      await computeUser(user.userId);
      processed += 1;
    } catch (error) {
      logger.error({ err: error, userId: user.userId }, 'analytics: compute failed for user');
      errors += 1;
    }
  }

  logger.info({ processed, errors }, 'analytics: compute done');
  return { processed, errors };
}
